daemons/ipa-kdb/ipa_kdb_principals.c

```diff
@@ -338,25 +337,84 @@ static void ipadb_validate_otp(struct ipadb_context *ipactx,
+    }
 
     /* Fetch the active token list. */
     kerr = ipadb_simple_search(ipactx, ipactx->base, LDAP_SCOPE_SUBTREE,
                                filter, (char**) attrs, &res);
     free(filter);
+    filter = NULL;
     if (kerr != 0 || res == NULL)
         return;
 
     /* Count the number of active tokens. */
     count = ldap_count_entries(ipactx->lcontext, res);
     ldap_msgfree(res);
 
-    /* If the user is configured for OTP, but has no active tokens, remove
-     * OTP from the list since the user obviously can't log in this way. */
-    if (count == 0)
-        *ua &= ~IPADB_USER_AUTH_OTP;
+    /*
+     * If there are no valid tokens then we need to remove the OTP flag,
+     * unless OTP is the only auth type allowed...
+     */
+    if (count == 0) {
+        if (*ua == IPADB_USER_AUTH_OTP) {
+            static const char *expired_ftmpl = "(&"
+                "(objectClass=ipaToken)(ipatokenOwner=%s)"
+                "(|(ipatokenNotAfter<=%s)(!(ipatokenNotAfter=*)))"
+            ")";
+
+            /* Remove the OTP flag for now */
+            *ua &= ~IPADB_USER_AUTH_OTP;
+
+            /*
+             * Ok, we "only" allow OTP, so if there is an expired token then
+             * add back the OTP flag as the server will double check the
+             * validity and reject the entire bind. Otherwise, this is the
+             * first time the user is authenticating and the user should be
+             * allowed to bind using its password
+             */
+            count = asprintf(&filter, expired_ftmpl, dn, datetime);
+            if (count < 0) {
+                ldap_memfree(dn);
+                return;
+            }
+
+            krb5_klog_syslog(LOG_INFO,
+                "Entry (%s) does not have a valid token and only OTP "
+                "authentication is supported, checking for expired tokens...",
+                dn);
+
+            kerr = ipadb_simple_search(ipactx, ipactx->base, LDAP_SCOPE_SUBTREE,
+                                       filter, (char**) attrs, &res);
+            free(filter);
+            if (kerr != 0 || res == NULL) {
+                ldap_memfree(dn);
+                return;
+            }
+
+            count = ldap_count_entries(ipactx->lcontext, res);
+            ldap_msgfree(res);
+            if (count > 0) {
+                /*
+                 * Ok we only allow OTP, and there are expired tokens so add
+                 * the OTP flag back, and the server will reject the bind.
+                 */
+                krb5_klog_syslog(LOG_INFO,
+                    "Entry (%s) does have an expired token so this user can "
+                    "not fall through to password auth", dn);
+                *ua |= IPADB_USER_AUTH_OTP;
+            }
+        } else {
+            /*
+             * Remove the OTP auth flag since there are no tokens and there
+             * are other authentication types we can use
+             */
+            *ua &= ~IPADB_USER_AUTH_OTP;
```

Comment:
Is this necessary? That should be the current state because it was set on line 369. I think retaining the block comment is a good idea to explain the fall-thru.

Reply:
I have been suggesting the following to @mreynolds389 in offline review:

One change I'd suggest is to avoid modifying *ua similarly in both branches. Instead, clear out OTP bit right inside the `(count == 0)` block and then only restore it if the `(count > 0)` is true:
```
if (count == 0) {
     *ua &= IPADB_USER_AUTH_OTP;
     if (*ua == 0) {

         check expired stuff, recalculate `count`

         if (count > 0) {
             *ua &= IPADB_USER_AUTH_OTP;
         }
     }
}
```
This way it would be more readable, in my opinion.